### DISCUSSION TITLE
DI configuration

### DISCUSSION THREAD

**Discussion by mnabe** [LuckyPennySoftware/AutoMapper#3700] [NA]
Can be found here: https://docs.automapper.org/en/latest/Getting-started.html

In the guide, the following setup example is given:

```c#
var config = new MapperConfiguration(cfg => cfg.CreateMap<Order, OrderDto>());

var mapper = config.CreateMapper();
// or
var mapper = new Mapper(config);
OrderDto dto = mapper.Map<OrderDto>(order);
```

However, in order to use dependency injection and inject the mapper in the classes where you want to use it, for example with constructor injection, you need to register the mapper with service container and a service lifetime, e.g.:

```c#
services.AddSingleton(mapper);
```

Removing the line `services.AddSingleton(mapper);` would result in this run-time error:

    InvalidOperationException: Unable to resolve service for type 'AutoMapper.IMapper' while attempting to activate {service}

So I am wondering what is the reason, why registering the mapper with the service container is not included in the getting started guide? I am sure, that it is me who misunderstands something, so I would love to hear explanations. Best regards and thanks for an awesome library.

**Comment by Tasteful** [comment]
If you scroll down to the https://docs.automapper.org/en/latest/Getting-started.html#where-do-i-configure-automapper section you can find a link to https://docs.automapper.org/en/latest/Dependency-injection.html#asp-net-core that describe how to configure AM with dependency injection.

**Reply by mnabe** [reply]
Thank you @Tasteful. I had already visited that page, but I guess I was too focused on looking explicitly for configuration options such as `services.AddSingleton(mapper);`, that the information didn't really sink in at first.

So it seems that it was never intended to be configured in the way, that I have usually done it. However, I think that I am not the only person, who has made that mistake; multiple SO/forum posts suggest to configure AutoMapper in the way, that I used as an example, being the most popular suggestion in several of them. For example top answers in these:

- https://stackoverflow.com/questions/40275195/how-to-set-up-automapper-in-asp-net-core
- https://stackoverflow.com/questions/54239669/asp-net-core-2-2-unable-to-resolve-service-for-type-automapper-imapper

I will make sure to refer to the documentation going forward.

Again, thanks for the help, much appreciated.
